kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
-    field @"deferred" .= indep
-
-    -- Push the dependent deferred pairs to the front of the queue.
-    traverse_ push dep
-    -- Apply the substitution to the queued pairs.
-    field @"queued" . transformQueue Lens.mapped %= substitute2
-
-    -- Apply the substitution to the accumulated matching solution.
-    field @"substitution" . Lens.mapped %= substitute1
-    field @"predicate" %= MultiAnd.map (Substitute.substitute subst)
-
-    return ()
-  where
-    variable = inject sVariable
-    Variable{variableName} = variable
-    isIndependent = not . any (hasFreeVariable variableName)
-    subst = Map.singleton variableName termLike
-    substitute2 :: Constraint variable -> Constraint variable
-    substitute2 (Constraint pair) = Constraint $ fmap substitute1 pair
-    substitute1 = substituteTermLike subst
-
-transformQueue ::
-    Functor f =>
-    Ord a =>
-    ((a -> f a) -> [a] -> f [a]) ->
-    (a -> f a) ->
-    Set a ->
-    f (Set a)
-transformQueue trans f x = Set.fromList <$> trans f (Set.toAscList x)
-
-substituteTermLike ::
-    MatchingVariable variable =>
-    Map (SomeVariableName variable) (TermLike variable) ->
-    TermLike variable ->
-    TermLike variable
-substituteTermLike subst = renormalizeBuiltins . Substitute.substitute subst
-
-occursCheck ::
-    (MatchingVariable variable, Monad simplifier) =>
-    SomeVariable variable ->
-    TermLike variable ->
-    MaybeT (MatcherT variable simplifier) ()
-occursCheck Variable{variableName} termLike =
-    (Monad.guard . not) (hasFreeVariable variableName termLike)
-
-definedTerm ::
-    MatchingVariable variable =>
-    MonadState (MatcherState variable) matcher =>
-    MonadReader (SideCondition variable) matcher =>
-    TermLike variable ->
-    matcher ()
-definedTerm termLike =
-    ask >>= definedTermWorker
-  where
-    definedTermWorker sideCondition
-        | SideCondition.isDefined sideCondition termLike =
-            return ()
-        | otherwise = field @"predicate" <>= definedTermLike
+        Simplifier (Either Text (MatchResult RewritingVariableName))
+    bind var term =
+        let varName = variableName var
+            freeVars = FreeVariables.toNames (freeVariables term)
+         in if not $ Set.disjoint freeVars boundSet
+                then failMatch "bound variable would escape binder"
+                else case Map.lookup varName subst of
+                    Nothing -> patternMatch' sideCondition rest deferred (isTermDefined var term) (Map.insert (variableName var) term subst)
+                    Just binding -> if binding == term then patternMatch' sideCondition rest deferred predicate subst else failMatch "nonlinear matching fails equality test"
+
+    isTermDefined ::
+        SomeVariable RewritingVariableName ->
+        TermLike RewritingVariableName ->
+        MultiAnd (Predicate RewritingVariableName)
+    isTermDefined var term
+        | SideCondition.isDefined sideCondition term || isSetVariable var = predicate
```

Comment:
Why `|| isSetVariable`?

Reply:
`isTermDefined` is a function intended to add a \ceil predicate for each element variable that gets added to the substitution. If it's a set variable, the ceil predicate is not needed because set variables are allowed to match \bottom.